### PR TITLE
99 model inference plots and loss curves

### LINKED ISSUE #99: Model inference plotting and loss curves for training, validation, and evaluation results
## Description:
Currently, model inference is being handled via the bash script: `bash scripts/models/inference_unet_with_mobious.bash` However, the workflow and configuration may be unclear or difficult to reproduce.

Similarly, the Python script:
`src/ready/apis/plot_losses.py -p <PATH> -lf1 <*.csv> -lf2 <*.csv>` only plots the loss values, and does not clearly show both training and validation losses which are essential for understanding model performance and diagnosing issues like overfitting.

This issue will address the following improvements:

## Definition of Done:
* [x]  Download models and document a reproducible workflow for cloning or fetching models from https://huggingface.co/mxochicale/ready_hf
* [x] Improve or rewrite the inference bash script to ensure the workflow is clear, reproducible, and well-documented.
* [x] Update or extend the plot_losses.py script to include:
  * [x] Combined plots showing both training and validation loss curves.
  * [x] Optionally, evaluation metrics (e.g. accuracy, Dice score) if available.
* [x] Add example plots to the documentation or README for visual reference.
* [x] Ensure the updated scripts are tested and paths can be passed via CLI arguments.
* [x] Document any new dependencies or usage examples.

### PR DESCRIPTION
**Pull request by stoppedtime24**
## Description
Fixes #99 : Changing plot_losses.py and plot_performance so that they can plot loss and performance for multiple files. Inference script now works on non-gpu machines (some plots missing because of no gpu). Updated docs to reflect changes.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### Suggested Checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation 
- [x] My changes generate no new warnings 
- [x] I have commented my code, particularly in hard-to-understand areas 
- [x] I have read the [`CONTRIBUTING`](https://github.com/UCL/ready/blob/main/CONTRIBUTING.md) docs
- [x] My code passes, following the style guidelines `pre-commit run -a` 
- [x] My code is properly tested with `pytest -sv tests`
- [x] I have requested review to this PR.
- [x] I have addressed and marked as resolved all the review comments in my PR.
- [x] Make sure your branch is up-to-date with main branch. See step 4 in [`CONTRIBUTING`](https://github.com/UCL/ready/blob/main/CONTRIBUTING.md#pull-request-pr-and-merge-to-main-branch).
- [x] Finally, I have selected `squash and merge`

### CODE REVIEW

File: data/mobious/models/README.md [hunk 1]
```diff
@@ -5,7 +5,10 @@ The models are hosted on Hugging Face at: https://huggingface.co/mxochicale/read
 You can either: Clone the repository to download all models and files, or Use wget to download individual files as needed.
 
 After downloading, make sure to move the models into their corresponding directory, following this format:
-`~/datasets/ready/mobious/models_a10080gb/{DATE_GPU_TYPE}` where `{DATE_GPU_TYPE}` refers to the training date and the GPU type used (e.g., 27-Jul-2025_NVIDIA_A100_80GB_PCI).
+`~/datasets/ready/mobious/models_a10080gb/{DATE_GPU_TYPE}` where `{DATE_GPU_TYPE}` refers to the training date and the GPU type used (e.g., 27-Jul-2025_NVIDIA_A100_80GB_PCI). This can be achieved using the following: `cp source/directory/ /destination/directory/`.
+
+Note: If you would like to use the model weights, ensure you download the actual from the Hugging Face repository. This can be accomplished by using the following:
+`wget https://huggingface.co/mxochicale/ready_hf/resolve/main/models/{model_folder}/{model_weights.pth}`. Otherwise, you will encounter an UnpicklingError.
```
Comment (mxochicale) [inline data/mobious/models/README.md:11]: Thanks for adding this note!

File: data/mobious/models/README.md [hunk 2]
```diff
@@ -42,16 +45,21 @@ bash scripts/models/inference_unet_with_mobious.bash
...
 
-* plot losses 
+* plot training and validation loss and performance metrics
 ```
-python src/ready/apis/plot_losses.py -p <PATH> -lf1 <*.csv> -lf2 <*.csv>
+# For losses
+python src/ready/apis/plot_losses.py -c <path/to/plot_losses.yml> 
+
+# For performance
+python src/ready/apis/plot_performance.py -c <path/to/plot_performance.yml>
```
Comment (mxochicale) [inline data/mobious/models/README.md:54]: Love these ones! thanks

File: data/mobious/models/README.md [hunk 3]
```diff
@@ -42,16 +45,21 @@ bash scripts/models/inference_unet_with_mobious.bash
...
 ```
 
-![fig](../../../docs/figs/losses_for_TRAINe100-80gbGPU.png)
+![fig](../../../docs/figs/Loss_Values_Multiple_Models.png)
+![fig](../../../docs/figs/Performance_Metrics_Multiple_Models.png)
```
Comment (mxochicale) [inline data/mobious/models/README.md:59]: Lovely plots!

File: src/ready/apis/plot_performance.py [hunk 4]
```diff
@@ -26,46 +27,53 @@
...
+        plt.bar(x_axis + (i - (n_bars -1 )/2) * width, current_validation_performance_values, width=width, label=current_validation_performance_label, alpha=0.5) 
+     
+    plt.xticks(x_axis, performance_metrics)
+    plt.ylabel('Values', fontsize=18)
+    plt.xlabel('Performance Metrics', fontsize=18)
+    plt.title('Training and Validation Performance Metrics', fontsize=18)
+    plt.legend(fontsize=15, loc='center right', framealpha=0.5)
+    plt.tick_params(axis='both', labelsize=17)
+    plt.grid(visible=True)
+    # plt.tight_layout()
```
Comment (mxochicale) [inline src/ready/apis/plot_performance.py:78]: Any difference with or without `plt.tight_layout()`? Maybe removed comment if there is no such difference

File: src/ready/apis/plot_losses.py [hunk 5]
```diff
@@ -19,42 +19,44 @@
...
     plt.xlabel("Epochs", fontsize=18)
     plt.ylabel("Loss", fontsize=18)
-    plt.tick_params(axis='both', labelsize=13)
-    # plt.legend(["naug_d1144", "waug_d1144", "waug_d0572", "waug_d0286", "waug_d0145"])
-    plt.legend(["naug_d1144", "waug_d1144", "waug_d0572"], fontsize=18)
-    plt.tight_layout()
+    plt.tick_params(axis='both', labelsize=13) 
+    plt.legend(framealpha=0.5)
     plt.grid()
+    # plt.tight_layout()
```
Comment (mxochicale) [inline src/ready/apis/plot_losses.py:61]: DITTO Any difference with or without plt.tight_layout()? Maybe removed comment if there is no such difference

File: src/ready/apis/inference_mobious.py [hunk 6]
```diff
@@ -100,7 +100,8 @@
     checkpoint_path = FULL_MODEL_PATH + '/' + str(model_name) + ".pth"
     model = UNet(nch_in=3, nch_out=4)
     model = model.to(device)
-    model.load_state_dict(torch.load(checkpoint_path, map_location=device))
+    print(checkpoint_path)
```
Comment (mxochicale) [inline src/ready/apis/inference_mobious.py:103]: would you like to use `logger.info(` instead of print?

**Review by mxochicale — APPROVED** [review]
Thanks @stoppedtime24 

Great PR, it looks good to me! I’ve just added a few comments, so please address them. Feel free to squash and merge whenever you’re happy with the changes.

Cheers, Miguel

File: src/ready/apis/plot_performance.py (hunk 4 above)
Comment (stoppedtime24) [reply src/ready/apis/plot_performance.py]: Will remove comment since using plt.tight_layout() ruins the plots.
